File: tools/public-qkey-converter/js/qkeyConverter.js
/*
 QKey Public Converter v1.0
 Official QKey format parser/generator for text, binary, JSON, CSV, XML
 Includes native blockchain/crypto type support
 No streaming API, no proprietary compression

 (c) 2025 Farrel Wilson — GPL-3.0-only
*/

export default class QKeyConverter {
    // --- QKey Text <-> JSON ---
    static parseText(qkeyText) {
        // Simple QKey text to JSON parser (similar to QKeyParser.simpleParse)
        const lines = qkeyText.split('\n');
        const result = { metadata: {}, schema: {}, records: [] };
        let currentRecord = {};
        let recordName = '';
        for (let line of lines) {
            line = line.trim();
            if (line.startsWith('//') || line === '') continue;
            if (line.includes(': {')) {
                recordName = line.split(':')[0].trim();
                currentRecord = {};
            } else if (line === '}' && recordName) {
                if (Object.keys(currentRecord).length > 0) {
                    result.records.push({ ...currentRecord, _id: recordName });
                }
                currentRecord = {};
                recordName = '';
            } else if (line.includes(':') && recordName) {
                const [key, ...valueParts] = line.split(':');
                let value = valueParts.join(':').trim();
                value = value.replace(/[",]/g, '').trim();
                currentRecord[key.trim()] = value;
            }
        }
        return result;
    }
    static stringifyText(jsonObj) {
        // JSON to QKey text generator
        let text = '';
        if (jsonObj.metadata && Object.keys(jsonObj.metadata).length > 0) {
            text += `@metadata: {\n`;
            for (const [k, v] of Object.entries(jsonObj.metadata)) {
                text += `  ${k}: "${v}"\n`;
            }
            text += `}\n`;
        }
        if (jsonObj.schema && Object.keys(jsonObj.schema).length > 0) {
            text += `schema: {\n`;
            for (const [k, v] of Object.entries(jsonObj.schema)) {
                text += `  ${k}: "${v}"\n`;
            }
            text += `}\n`;
        }
        if (jsonObj.records) {
            for (const rec of jsonObj.records) {
                const recName = rec._id || 'record';
                text += `${recName}: {\n`;
                for (const [k, v] of Object.entries(rec)) {
                    if (k === '_id') continue;
                    text += `  ${k}: "${v}"\n`;
                }
                text += `}\n`;
            }
        }
        return text;
    }

    // --- QKB Binary <-> JSON ---
    static parseBinary(input) {
        // Accept Buffer | Uint8Array | ArrayBuffer | string (UTF-8 encoded QKey text)
        try {
            let text = '';
            const hasBuffer = (typeof Buffer !== 'undefined' && typeof Buffer.from === 'function' && typeof Buffer.isBuffer === 'function');
            if (hasBuffer && Buffer.isBuffer(input)) {
                text = input.toString('utf8');
            } else if (typeof input === 'string') {
                text = input;
            } else if (typeof TextDecoder !== 'undefined' && (input instanceof Uint8Array || input instanceof ArrayBuffer)) {
                const u8 = input instanceof Uint8Array ? input : new Uint8Array(input);
                text = new TextDecoder('utf-8').decode(u8);
            } else {
                return { metadata: {}, schema: {}, records: [] };
            }
            return QKeyConverter.parseText(text);
        } catch (_) {
            return { metadata: {}, schema: {}, records: [] };
        }
    }
    static stringifyBinary(jsonObj) {
        // Return Buffer in Node, Uint8Array in browser
        const text = QKeyConverter.stringifyText(jsonObj);
        const hasBuffer = (typeof Buffer !== 'undefined' && typeof Buffer.from === 'function');
        if (hasBuffer) return Buffer.from(text, 'utf8');
        if (typeof TextEncoder !== 'undefined') return new TextEncoder().encode(text);
        // Final fallback: return plain string (caller should handle)
        return text;
    }

    // --- CSV <-> QKey/JSON ---
    static parseCSV(csvText) {
        // CSV with basic RFC4180 handling: quotes, commas, newlines inside quotes
        const lines = csvText.replace(/\r\n?/g, '\n').split('\n');
        if (lines.length === 0) return { records: [] };
        const parseLine = (line) => {
            const out = [];
            let cur = '';
            let inQuotes = false;
            for (let i = 0; i < line.length; i++) {
                const ch = line[i];
                if (ch === '"') {
                    if (inQuotes && line[i + 1] === '"') { cur += '"'; i++; }
                    else inQuotes = !inQuotes;
                } else if (ch === ',' && !inQuotes) {
                    out.push(cur);
                    cur = '';
                } else {
                    cur += ch;
                }
            }
            out.push(cur);
            return out.map(s => s.trim());
        };
        // Accumulate possibly multi-line quoted records
        const rows = [];
        let buffer = '';
        let quoteCount = 0;
        for (const raw of lines) {
            const line = buffer ? (buffer + '\n' + raw) : raw;
            const qc = (line.match(/"/g) || []).length;
            quoteCount += qc;
            if (quoteCount % 2 === 0) {
                rows.push(line);
                buffer = '';
                quoteCount = 0;
            } else {
                buffer = line;
            }
        }
        if (buffer) rows.push(buffer);
        if (rows.length === 0) return { records: [] };
        const headers = parseLine(rows[0]);
        const records = rows.slice(1).filter(r => r.trim() !== '').map(row => {
            const values = parseLine(row);
            const rec = {};
            headers.forEach((h, i) => rec[h] = (values[i] ?? '').trim());
            return rec;
        });
        return { records };
    }
    static stringifyCSV(jsonObj) {
        // JSON/QKey to CSV generator with quoting when needed
        if (!jsonObj.records || jsonObj.records.length === 0) return '';
        const headers = Object.keys(jsonObj.records[0]).filter(k => k !== '_id');
        const esc = (v) => {
            const s = (v == null ? '' : String(v));
            if (/[,"\n]/.test(s)) return '"' + s.replace(/"/g, '""') + '"';
            return s;
        };
        let csv = headers.join(',') + '\n';
        for (const rec of jsonObj.records) {
            csv += headers.map(h => esc(rec[h])).join(',') + '\n';
        }
        return csv;
    }

    // --- XML <-> QKey/JSON ---
    static parseXML(xmlText) {
        // Prefer DOM parsing in browsers; fallback to simple tag parsing
        try {
            if (typeof DOMParser !== 'undefined') {
                const doc = new DOMParser().parseFromString(xmlText, 'application/xml');
                const recNodes = doc.getElementsByTagName('record');
                const records = [];
                for (let i = 0; i < recNodes.length; i++) {
                    const rec = {};
                    const children = recNodes[i].children;
                    for (let j = 0; j < children.length; j++) {
                        rec[children[j].tagName] = (children[j].textContent || '').trim();
                    }
                    records.push(rec);
                }
                return { records };
            }
        } catch (_) { /* fall through */ }
        // Fallback: basic tag parsing
        const records = [];
        const recRegex = /<record>([\s\S]*?)<\/record>/g;
        let match;
        while ((match = recRegex.exec(xmlText)) !== null) {
            const rec = {};
            const fieldRegex = /<([a-zA-Z0-9_]+)>([\s\S]*?)<\/\1>/g;
            let fieldMatch;
            while ((fieldMatch = fieldRegex.exec(match[1])) !== null) {
                rec[fieldMatch[1]] = fieldMatch[2].trim();
            }
            records.push(rec);
        }
        return { records };
    }
    static stringifyXML(jsonObj) {
        // JSON/QKey to XML generator
        if (!jsonObj.records || jsonObj.records.length === 0) return '<records></records>';
        let xml = '<records>\n';
        for (const rec of jsonObj.records) {
            xml += '  <record>\n';
            for (const [k, v] of Object.entries(rec)) {
                if (k === '_id') continue;
                xml += `    <${k}>${v}</${k}>\n`;
            }
            xml += '  </record>\n';
        }
        xml += '</records>';
        return xml;
    }

    // --- Blockchain/Crypto Types ---
    static validateEthereumAddress(address) {
        // Basic Ethereum address validation
        return /^0x[a-fA-F0-9]{40}$/.test(address);
    }
    static validateBitcoinAddress(address) {
        // Legacy Base58 (simplified) or Bech32 (bc1/tb1)
        if (/^[13][a-km-zA-HJ-NP-Z1-9]{25,34}$/.test(address)) return true;
        if (/^(bc1|tb1)[ac-hj-np-z02-9]{11,71}$/.test(address)) return true;
        return false;
    }
    static validateSignature(signature) {
        // Basic signature format validation (trimmed)
        return typeof signature === 'string' && signature.trim().length > 0;
    }
    static isHex(str) {
        return /^[0-9a-fA-F]+$/.test(str);
    }
    static isBase64(str) {
        return /^(?:[A-Za-z0-9+\/]{4})*(?:[A-Za-z0-9+\/]{2}==|[A-Za-z0-9+\/]{3}=)?$/.test(str);
    }
    static validateTxHash(hash) {
        // Ethereum tx hash: 0x + 64 hex chars
        return /^0x[a-fA-F0-9]{64}$/.test(hash);
    }
    static validateNFTTokenId(tokenId) {
        // NFT token id: numeric or hex
        return /^[0-9]+$/.test(tokenId) || QKeyConverter.isHex(tokenId);
    }

    // --- Commercial Feature Notifications ---
    static useAdvancedCompression() {
        console.warn("Advanced QKB compression is a commercial feature. Using basic compression instead.");
        console.info("For commercial licensing options, contact sales@Querykey.com");
        return false;
    }
    
    static useStreamingAPI() {
        console.warn("Streaming API is a commercial feature not available in the open source edition.");
        console.info("For commercial licensing options, contact sales@Querykey.com");
        return false;
    }
    
    static useEnterpriseSecurityFeatures() {
        console.warn("Enterprise security features are not available in the open source edition.");
        console.info("For commercial licensing options, contact sales@Querykey.com");
        return false;
    }
    
    static useHighPerformanceBinary() {
        console.warn("High-performance binary format is a commercial feature.");
        console.info("For commercial licensing options, contact sales@Querykey.com");
        return false;
    }

    // --- Validation ---
    static validateQKeyText(qkeyText) {
        // Basic QKey text format validation
        return typeof qkeyText === 'string' && qkeyText.includes(': {');
    }
    static validateQKB(qkbBuffer) {
        // Accept Buffer | Uint8Array | ArrayBuffer
        if (typeof qkbBuffer === 'undefined' || qkbBuffer === null) return false;
        const isArrayBuffer = (qkbBuffer instanceof ArrayBuffer);
        const isUint8 = (typeof Uint8Array !== 'undefined' && qkbBuffer instanceof Uint8Array);
        const isBuffer = (typeof Buffer !== 'undefined' && typeof Buffer.isBuffer === 'function' && Buffer.isBuffer(qkbBuffer));
        if (isArrayBuffer) return (qkbBuffer.byteLength || 0) > 0;
        if (isUint8) return qkbBuffer.length > 0;
        if (isBuffer) return qkbBuffer.length > 0;
        return false;
    }
}

File: tools/public-qkey-converter/python/qkey_converter.py
"""
QKey Public Converter v1.0 - Python Version
Official QKey format parser/generator for text, binary, JSON, CSV, XML
Includes native blockchain/crypto type support

Copyright (c) 2025 Farrel Wilson
License: GPL-3.0-only

This program is free software: you can redistribute it and/or modify
it under the terms of the GNU General Public License as published by
the Free Software Foundation, version 3.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program. If not, see <https://www.gnu.org/licenses/>.
"""

import json
import re
import csv
import io
import xml.etree.ElementTree as ET
from typing import Dict, List, Any, Union, Optional


class QKeyConverter:
    """Official QKey format converter with blockchain/crypto support"""
    
    @staticmethod
    def parse_text(qkey_text: str) -> Dict[str, Any]:
        """Parse QKey text to JSON/dict"""
        lines = qkey_text.split('\n')
        result = {'metadata': {}, 'schema': {}, 'records': []}
        current_record = {}
        record_name = ''
        
        for line in lines:
            line = line.strip()
            if line.startswith('//') or line == '':
                continue
                
            if ': {' in line:
                record_name = line.split(':')[0].strip()
                current_record = {}
            elif line == '}' and record_name:
                if current_record:
                    current_record['_id'] = record_name
                    result['records'].append(current_record)
                current_record = {}
                record_name = ''
            elif ':' in line and record_name:
                parts = line.split(':', 1)
                key = parts[0].strip()
                value = parts[1].strip().strip('"').strip(',')
                current_record[key] = value
                
        return result
    
    @staticmethod
    def stringify_text(json_obj: Dict[str, Any]) -> str:
        """Convert JSON/dict to QKey text"""
        text = ''
        
        # Metadata section
        if 'metadata' in json_obj and json_obj['metadata']:
            text += '@metadata: {\n'
            for k, v in json_obj['metadata'].items():
                text += f'  {k}: "{v}"\n'
            text += '}\n'
        
        # Schema section
        if 'schema' in json_obj and json_obj['schema']:
            text += 'schema: {\n'
            for k, v in json_obj['schema'].items():
                text += f'  {k}: "{v}"\n'
            text += '}\n'
        
        # Records section
        if 'records' in json_obj and json_obj['records']:
            for record in json_obj['records']:
                record_name = record.get('_id', 'record')
                text += f'{record_name}: {{\n'
                for k, v in record.items():
                    if k != '_id':
                        text += f'  {k}: "{v}"\n'
                text += '}\n'
        
        return text
    
    @staticmethod
    def parse_binary(qkb_buffer: bytes) -> Dict[str, Any]:
        """Parse QKB binary to JSON/dict (basic UTF-8 implementation)"""
        try:
            text = qkb_buffer.decode('utf-8')
            return QKeyConverter.parse_text(text)
        except:
            return {'metadata': {}, 'schema': {}, 'records': []}
    
    @staticmethod
    def stringify_binary(json_obj: Dict[str, Any]) -> bytes:
        """Convert JSON/dict to QKB binary (basic UTF-8 implementation)"""
        text = QKeyConverter.stringify_text(json_obj)
        return text.encode('utf-8')
    
    @staticmethod
    def parse_csv(csv_text: str) -> Dict[str, Any]:
        """Parse CSV to JSON/dict"""
        reader = csv.DictReader(io.StringIO(csv_text))
        records = []
        for row in reader:
            # Clean up the row data
            clean_row = {k.strip(): v.strip() if v else '' for k, v in row.items()}
            records.append(clean_row)
        return {'records': records}
    
    @staticmethod
    def stringify_csv(json_obj: Dict[str, Any]) -> str:
        """Convert JSON/dict to CSV"""
        if not json_obj.get('records'):
            return ''
        
        records = json_obj['records']
        if not records:
            return ''
        
        # Get headers (exclude _id)
        headers = [k for k in records[0].keys() if k != '_id']
        
        output = io.StringIO()
        writer = csv.DictWriter(output, fieldnames=headers)
        writer.writeheader()
        
        for record in records:
            row = {k: record.get(k, '') for k in headers}
            writer.writerow(row)
        
        return output.getvalue()
    
    @staticmethod
    def parse_xml(xml_text: str) -> Dict[str, Any]:
        """Parse XML to JSON/dict (basic implementation)"""
        try:
            root = ET.fromstring(xml_text)
            records = []
            
            for record_elem in root.findall('.//record'):
                record = {}
                for child in record_elem:
                    record[child.tag] = child.text or ''
                records.append(record)
            
            return {'records': records}
        except:
            return {'records': []}
    
    @staticmethod
    def stringify_xml(json_obj: Dict[str, Any]) -> str:
        """Convert JSON/dict to XML"""
        if not json_obj.get('records'):
            return '<records></records>'
        
        root = ET.Element('records')
        
        for record in json_obj['records']:
            record_elem = ET.SubElement(root, 'record')
            for k, v in record.items():
                if k != '_id':
                    child = ET.SubElement(record_elem, k)
                    child.text = str(v)
        
        return ET.tostring(root, encoding='unicode')
    
    # Blockchain/Crypto Utilities
    @staticmethod
    def validate_ethereum_address(address: str) -> bool:
        """Validate Ethereum address format (basic format check)."""
        if not isinstance(address, str):
            return False
        return bool(re.match(r'^0x[0-9a-fA-F]{40}$', address))
    
    @staticmethod
    def validate_bitcoin_address(address: str) -> bool:
        """Validate Bitcoin address format (legacy and Bech32)."""
        if not isinstance(address, str):
            return False
        # Legacy Base58 (very simplified length/range)
        if re.match(r'^[13][a-km-zA-HJ-NP-Z1-9]{25,34}$', address):
            return True
        # Bech32 (bc1... or tb1...)
        if re.match(r'^(bc1|tb1)[ac-hj-np-z02-9]{11,71}$', address):
            return True
        return False
    
    @staticmethod
    def validate_signature(signature: str) -> bool:
        """Validate signature format (basic sanity check)."""
        return isinstance(signature, str) and len(signature.strip()) > 0
    
    @staticmethod
    def is_hex(text: str) -> bool:
        """Check if string is hexadecimal"""
        return bool(re.match(r'^[0-9a-fA-F]+$', text))
    
    @staticmethod
    def is_base64(text: str) -> bool:
        """Check if string is Base64"""
        import base64
        try:
            base64.b64decode(text, validate=True)
            return True
        except:
            return False
    
    @staticmethod
    def validate_tx_hash(tx_hash: str) -> bool:
        """Validate transaction hash format"""
        return bool(re.match(r'^0x[a-fA-F0-9]{64}$', tx_hash))
    
    @staticmethod
    def validate_nft_token_id(token_id: str) -> bool:
        """Validate NFT token ID format"""
        return token_id.isdigit() or QKeyConverter.is_hex(token_id)
    
    # Validation
    @staticmethod
    def validate_qkey_text(qkey_text: str) -> bool:
        """Validate QKey text format"""
        return isinstance(qkey_text, str) and ': {' in qkey_text
    
    @staticmethod
    def validate_qkb(qkb_buffer: bytes) -> bool:
        """Validate QKB binary format"""
        return isinstance(qkb_buffer, bytes) and len(qkb_buffer) > 0
    
    # --- Commercial Feature Notifications ---
    @staticmethod
    def use_advanced_compression() -> bool:
        """Alert about commercial feature: Advanced QKB compression"""
        print("Warning: Advanced QKB compression is a commercial feature. Using basic compression instead.")
        print("For commercial licensing options, contact sales@Querykey.com")
        return False
    
    @staticmethod
    def use_streaming_api() -> bool:
        """Alert about commercial feature: Streaming API"""
        print("Warning: Streaming API is a commercial feature not available in the open source edition.")
        print("For commercial licensing options, contact sales@Querykey.com")
        return False
    
    @staticmethod
    def use_enterprise_security_features() -> bool:
        """Alert about commercial feature: Enterprise security features"""
        print("Warning: Enterprise security features are not available in the open source edition.")
        print("For commercial licensing options, contact sales@Querykey.com")
        return False
    
    @staticmethod
    def use_high_performance_binary() -> bool:
        """Alert about commercial feature: High-performance binary format"""
        print("Warning: High-performance binary format is a commercial feature.")
        print("For commercial licensing options, contact sales@Querykey.com")
        return False


# Example usage
if __name__ == '__main__':
    # Test basic functionality
    qkey_text = """
user1: {
  name: "Alice"
  email: "alice@example.com"
  wallet: "0x1234567890123456789012345678901234567890"
}
"""
    
    print("=== QKey Python Converter Test ===")
    
    # Parse QKey text
    data = QKeyConverter.parse_text(qkey_text)
    print("Parsed data:", json.dumps(data, indent=2))
    
    # Generate QKey text
    generated = QKeyConverter.stringify_text(data)
    print("Generated QKey text:")
    print(generated)
    
    # Test blockchain validation
    eth_addr = "0x1234567890123456789012345678901234567890"
    print(f"Ethereum address valid: {QKeyConverter.validate_ethereum_address(eth_addr)}")
    
    print("=== Test Complete ===")
